### ISSUE TITLE
Własne ID klienta -> numer konta bankowego

### ISSUE THREAD

**Issue by Mac1ej** [lmsgit/lms#1426]
Witam,
zastanawiam się nad możliwością dodawania klienta z ustalonym wcześniej ID (np. nadany w innym systemie). Istnieje wprawdzie pole "Id zewnętrzne", ale nie jest ono traktowane tak jak ID do generowania numeru rachunku bankowego (do masowego przetwarzania transakcji).
Zatem co w przypadku chęci dopisania klienta z konkretnym ID, aby miał konkretny numer rachunku bankowego?
Najszybsza opcja, jaka przychodzi mi do głowy, do dopisanie takich klientów "ręcznie" do bazy danych (lub wpisanie przez LMS i modyfikacja jedynie ich ID w bazie). Oczywiście z założenia ID będzie nadal unikalne. W tym wypadku pozostaną niewykorzystane numery ID (LMS numeruje po kolei od 1) pomiędzy klientami, gdy w innym systemie numeracja nie jest po kolei, bo np. numer ID to numer PESEL klienta.
Druga opcja, zapewne lepsza pod kątem spójności bazy danych, to używanie "Id zewnętrznego", a LMS swoją drogą niech numeruje sobie klientów po po kolei, ale wtedy należy dodać opcję umożliwiającą generowanie rachunków bankowych z wykorzystaniem "Id zewnętrznego" zamiast "ID". Co o tym myślicie?
Może jakaś zmienna w konfiguracji interfejsu użytkownika typu true/false odpowiadająca za używanie "Id zewnętrznego" zamiast "ID"... być może nie tylko do generowania rachunków bankowych, ale również do innych rzeczy - logowania do BOK, wyświetlania na fakturach itp. Po co klient ma zapamiętywać kolejne swoje numery identyfikacyjne, jak każdy i tak ma już swój PESEL? Wiadomo, że dla firm/instytucji będzie inaczej (chociaż można dać np. Regon), ale dla konsumentów to mniej zamieszania :)
Pozdrawiam
Maciej Aneszko

**Comment by interduo** [comment]
Jaki problem rozwiązuje dodanie klienta z ustalonym wcześniej konkretnym id?

**Comment by Mac1ej** [comment]
Będzie miał numer rachunku bankowego taki, jak powinien mieć (ustalony np. z wykorzystaniem numeru PESEL), a nie na podstawie kolejnego numeru.

**Comment by interduo** [comment]
Czy masz w swoim banku dostępnych ostatnuch 11 cyfr do rachunków masowych?

Co daje nr PESEL w rachunku? Co z firmami?

Z tego co pamiętam u klienta da się nadpisać ręcznie numer rachunku.

Dlaczego numery rachunków bankowych następujące po sobie są złe?

**Comment by interduo** [comment]
Zauważ to, że i tak tego numeru nie będzie pamiętał ponieważ (chyba na początku) numeru konta bankowego w formacie IBAN jest suma kontrolna.

**Comment by Mac1ej** [comment]
Tak, mam w banku 11 cyfr dla identyfikatora kontrahenta (11 ostatnich cyfr w indywidualnym numerze rachunku bankowego).

PESEL dotyczy konsumentów. NIP czy Regon są krótsze (można użyć dla firm).

Tak, da się wpisać klientowi numer rachunku bankowego ręcznie, ale chyba nie o to chodzi... równie dobrze mogę faktury również wystawiać ręcznie.
Poza tym na fakturach (wystawionych w LMS) i tak pojawia się numer rachunku wygenerowany z prefiksu konta bankowego i ID klienta (zapewne można edytować szablon faktury... ale czy o to chodzi?). Fajnie by było móc wykorzystać mechanizm LMS, który z ID generuje numer rachunku bankowego... nie zmuszając przy tym, aby ID było kolejną liczbą.

Numery rachunków bankowych następujące po sobie nie są złe, o ile podpisujemy nowe umowy i zgadzamy się na to, że klient będzie posiadał kolejny numer ID (mimo że nie jest to potrzebne, bo już i tak posiada unikalny numer PESEL).
Gorzej ma się sprawa z obecnymi klientami, którzy w umowach mają wpisane indywidualne numery rachunków - trudno ich wtedy przepisać do LMS-a, a przecież w tym celu nie będę zmieniać klientom  numerów kont bankowych. Jeśli ktoś nie używał wcześniej LMS-a i chciałby zacząć, a ma już klientów z nadanymi numerami kont bankowych, to może natrafić na taki właśnie problem (w sumie dosyć prosty do rozwiązania - opcja, o której wspomniałem, to raczej niedużo kodu).

**Comment by Mac1ej** [comment]
> Zauważ to, że i tak tego numeru nie będzie pamiętał ponieważ (chyba na początku) numeru konta bankowego w formacie IBAN jest suma kontrolna.

Tak, na początku jest suma kontrolna, którą LMS ładnie wylicza :)
Nie biorę pod uwagę tego, że klient będzie zapamiętywał numer rachunku bankowego - od tego ma fakturę i umowę, aby sobie sprawdzić. Nawet bez sumy kontrolnej, zakładając, że klient pamięta już swój PESEL, jest sporo cyfr do zapamiętania (bank, identyfikator usługi). Ale przynajmniej w łatwy sposób może sprawdzić, że na pewno wpłaca na swoje konto.

Chodzi przede wszystkim o to, że jeśli ktoś już wygenerował swoim klientom takie rachunki (czy to na podstawie numeru PESEL, numeru umowy czy numeru klienta - nie ma większego znaczenia), trudno te numery zgrać z kolejnymi numerami ID przy wpisywaniu klientów do LMS.

**Comment by interduo** [comment]
Czy na prawdę LMS nie wrzuca numeru bankowego ustawionego ręcznie na fakturę ustawioną automatycznie?

**Comment by Mac1ej** [comment]
Zrobiłem kilka testów i niestety ani razu nie udało mi się uzyskać faktury z numerem rachunku zapisanym w polu "Alternatywne rachunki bankowe" danego klienta.
Oczywiście zaznaczyłem checkbox "faktury", aby ten numer pojawiał się na fakturach (zgodnie z podpowiedzią z dymka), ale to nie pomogło. Zresztą przy wyłączonym checkboxie jest to samo.
Faktury sprawdzałem zarówno przy generowaniu ich do html (szablon invoice.html) jak i do pdf (szablon standard).
Generowałem je ręcznie (?m=invoicenew) jak i za pomocą lms-payments.php. Efekt ten sam.
Gdy w danych firmy (oddziału) wpisany jest prefiks konta bankowego do płatności masowych, to się na fakturach pojawia własnie nr konta wygenerowany na podstawie tego prefiksu i ID klienta (suma kontrolna obliczona jest poprawnie). Gdy ten prefiks skasuję z danych firmy (oddziału), to faktura generuje się z pustą zawartością "Konto bankowe:".
Niestety ani razu nie pojawił się ten alternatywny rachunek bankowy przypisany klientowi.
Czy ktoś używa tej opcji i mu ona działa?

**Comment by chilek** [comment]
Spróbuj ustawić w konfiguracji **invoices.show_all_accounts=true**.
